library/core/src/iter/adapters/array_chunks.rs

```diff
@@ -0,0 +1,232 @@
+use core::iter::FusedIterator;
+use core::mem::{self, MaybeUninit};
+use core::ops::{ControlFlow, Try};
+use core::ptr;
+
+#[derive(Debug)]
+struct Buffer<T, const N: usize> {
+    array: [MaybeUninit<T>; N],
+    init: usize,
+}
+
+impl<T, const N: usize> Buffer<T, N> {
+    fn new() -> Self {
+        Self { array: MaybeUninit::uninit_array(), init: 0 }
+    }
+}
+
+impl<T: Clone, const N: usize> Clone for Buffer<T, N> {
+    fn clone(&self) -> Self {
+        let mut new = Self::new();
+        // SAFETY: this raw slice contains only the initialized objects.
+        let src = unsafe { MaybeUninit::slice_assume_init_ref(&self.array[..self.init]) };
+        MaybeUninit::write_slice_cloned(&mut new.array[..self.init], src);
+        new.init = self.init;
+        new
+    }
+}
+
+impl<T, const N: usize> Drop for Buffer<T, N> {
+    fn drop(&mut self) {
+        debug_assert!(self.init <= N);
+
+        let initialized_part = &mut self.array[..self.init];
+
+        // SAFETY: this raw slice will contain only the initialized objects
+        // that have not been dropped or moved.
+        unsafe {
+            ptr::drop_in_place(MaybeUninit::slice_assume_init_mut(initialized_part));
+        }
+    }
+}
+
+// FIXME: Combine with `Guard` in `collect_into_array`.
+struct Guard<T, const N: usize> {
+    ptr: *mut T,
+    init: usize,
+}
+
+impl<T, const N: usize> Drop for Guard<T, N> {
+    fn drop(&mut self) {
+        debug_assert!(self.init <= N);
+
+        let initialized_part = crate::ptr::slice_from_raw_parts_mut(self.ptr, self.init);
+
+        // SAFETY: this raw slice will contain only initialized objects.
+        unsafe {
+            crate::ptr::drop_in_place(initialized_part);
+        }
+    }
+}
+
+impl<T, const N: usize> Guard<T, N> {
+    fn new(array: &mut [MaybeUninit<T>; N]) -> Self {
+        Self { ptr: MaybeUninit::slice_as_mut_ptr(array), init: 0 }
+    }
+
+    fn with<R, F>(buffer: &mut Buffer<T, N>, f: F) -> R
+    where
+        F: FnOnce(&mut [MaybeUninit<T>; N], &mut usize) -> R,
+    {
+        let mut array = MaybeUninit::uninit_array();
+        let mut guard = Self::new(&mut array);
+        if buffer.init > 0 {
+            array = mem::replace(&mut buffer.array, MaybeUninit::uninit_array());
+            guard.init = mem::replace(&mut buffer.init, 0);
+        }
+        let res = f(&mut array, &mut guard.init);
+        if guard.init > 0 {
```

Comment:
Similar here, what's the advantage of the extra compare here? Wouldn't it be better to tie moving the data into the adapter to the `next == None` condition?